README.md
# Progresso

A linear progress bar for Flutter.

## Getting Started

Add to pubspec.yaml 

```yaml
progresso: <lastest version>
```

import in your flutter project

```dart
import 'package:progresso/progresso.dart';
```

## Features

In its simplest form Progresso takes a progress parameter (0.0 to 1.0)

![basic](https://raw.githubusercontent.com/russbiggs/progresso/main/imgs/basic.png)

```dart
Progresso(progress: 0.5);
```

You can also configure the color

![custom color](https://raw.githubusercontent.com/russbiggs/progresso/main/imgs/color.png)


```dart
Progresso(
    progress: 0.5, 
    progressColor: Colors.red, 
    backgroundColor: Colors.blue
);
```

Progresso can even start at a value that is not zero!

![non-zero start](https://raw.githubusercontent.com/russbiggs/progresso/main/imgs/start.png)


```dart 
Progresso(start: 0.3, progress: 0.5);
```

Give the progress bar different stroke caps:

![round end caps](https://raw.githubusercontent.com/russbiggs/progresso/main/imgs/endcap.png)


```dart 
Progresso(
    progress: 0.5, 
    progressStrokeCap: StrokeCap.round,
    backgroundStrokeCap: StrokeCap.round
);
```

Give a List of points to show points along the progress bar

![points](https://raw.githubusercontent.com/russbiggs/progresso/main/imgs/points.png)


```dart
Progresso(progress: 0.5, points: [0.2, 0.4]);
```

CHANGELOG.md
## [0.1.0] - February 28 2021

* initial release

## [0.1.1] - March 1 2021

* update readme
* add example 
* add doc strings

## [0.2.0] - March 9 2021

* redraw points when data changes
* redraw background stroke cap when background stroke cap changes
* redraw progress stroke cap when progress stroke cap changes


## [0.3.0] - March 31 2021

* migrate to null safety

lib/progresso.dart
library progresso;

import 'package:flutter/material.dart';

///
class Progresso extends StatefulWidget {
  /// The starting position of the progress bar. Defaults to 0.0 must be less
  /// than 1.0
  final double start;

  /// The progress position of the progress bar. Defaults to 0.0 must be less
  /// than 1.0 and greater than start.
  final double progress;

  /// a List of double values that are represented on the progress line as
  /// points. Must be greater than or equal to 0.0 and less than or equal to 1.0
  final List<double> points;

  /// The Color of the progress bar
  final Color progressColor;

  /// The color of the background bar
  final Color backgroundColor;

  /// The stroke width of the progress bar
  final double progressStrokeWidth;

  /// The cap style of the progress bar. Defaults to square, options include
  /// StrokeCap.round, StrokeCap.square, StrokeCap.butt
  final StrokeCap progressStrokeCap;

  ///the stroke width of the background bar
  final double backgroundStrokeWidth;

  /// The cap style of the progress bar. Defaults to square, options include
  /// StrokeCap.round, StrokeCap.square, StrokeCap.butt
  final StrokeCap backgroundStrokeCap;

  /// The Color of the outer circle of the points given in the points parameter
  /// defaults to Colors.blue
  final Color pointColor;

  /// The Color of the outer circle of the points given in the points parameter
  /// defaults to Colors.white
  final Color pointInnerColor;

  /// The radius of the outer circle of the points given in the points parameter
  /// defaults to 7.5
  final double pointRadius;

  /// The radius of the inner circle of the points given in the points parameter
  /// defaults to 2.5
  final double pointInnerRadius;

  Progresso(
      {Key? key,
      this.start = 0.0,
      this.progress = 0.0,
      this.progressColor = Colors.blue,
      this.backgroundColor = Colors.grey,
      this.progressStrokeWidth = 10.0,
      this.backgroundStrokeWidth = 5.0,
      this.progressStrokeCap = StrokeCap.square,
      this.backgroundStrokeCap = StrokeCap.square,
      this.pointColor = Colors.blue,
      this.pointInnerColor = Colors.white,
      this.pointRadius = 7.5,
      this.pointInnerRadius = 2.5,
      this.points = const []})
      : super(key: key);

  @override
  _ProgressoState createState() => _ProgressoState();
}

class _ProgressoState extends State<Progresso> {
  @override
  Widget build(BuildContext context) {
    return CustomPaint(
        foregroundPainter: _ProgressoPainter(
            progress: widget.progress,
            start: widget.start,
            progressColor: widget.progressColor,
            backgroundColor: widget.backgroundColor,
            progressStrokeWidth: widget.progressStrokeWidth,
            progressStrokeCap: widget.progressStrokeCap,
            backgroundStrokeWidth: widget.backgroundStrokeWidth,
            backgroundStrokeCap: widget.backgroundStrokeCap,
            pointColor: widget.pointColor,
            pointInnerColor: widget.pointInnerColor,
            pointRadius: widget.pointRadius,
            pointInnerRadius: widget.pointInnerRadius,
            points: widget.points),
        child: Center());
  }
}

class _ProgressoPainter extends CustomPainter {
  final Paint _paintBackground = new Paint();
  final Paint _paintProgress = new Paint();
  final Paint _paintPoint = new Paint();
  final Paint _paintPointCenter = new Paint();
  final Color backgroundColor;
  final Color progressColor;
  final double start;
  final double progress;
  final double progressStrokeWidth;
  final StrokeCap progressStrokeCap;
  final double backgroundStrokeWidth;
  final StrokeCap backgroundStrokeCap;
  final Color pointColor;
  final Color pointInnerColor;
  final double pointRadius;
  final double pointInnerRadius;
  final List<double> points;

  _ProgressoPainter(
      {required this.start,
      required this.progress,
      required this.progressColor,
      required this.backgroundColor,
      required this.progressStrokeWidth,
      required this.progressStrokeCap,
      required this.backgroundStrokeWidth,
      required this.backgroundStrokeCap,
      required this.points,
      required this.pointColor,
      required this.pointInnerColor,
      required this.pointInnerRadius,
      required this.pointRadius}) {
    _paintBackground.color = backgroundColor;
    _paintBackground.style = PaintingStyle.stroke;
    _paintBackground.strokeCap = backgroundStrokeCap;
    _paintBackground.strokeWidth = backgroundStrokeWidth;
    _paintProgress.color = progressColor;
    _paintProgress.style = PaintingStyle.stroke;
    _paintProgress.strokeCap = progressStrokeCap;
    _paintProgress.strokeWidth = progressStrokeWidth;
    _paintPoint.color = pointColor;
    _paintPoint.style = PaintingStyle.fill;
    _paintPointCenter.color = pointInnerColor;
    _paintPointCenter.style = PaintingStyle.fill;
  }

  @override
  void paint(Canvas canvas, Size size) {
    final startOffset = Offset(0.0, size.height / 2);
    final endOffset = Offset(size.width, size.height / 2);
    canvas.drawLine(startOffset, endOffset, _paintBackground);
    final xStart = size.width * start;
    var cappedProgress = progress;
    if (progress > 1) {
      cappedProgress = 1.0;
    }
    var xProgress = size.width * cappedProgress;
    final progressStart = Offset(xStart, size.height / 2);
    canvas.drawLine(
        progressStart, Offset(xProgress, size.height / 2), _paintProgress);
    for (var point in points) {
      var pointPos = size.width * point;
      final pointX = Offset(pointPos, size.height / 2);
      canvas.drawCircle(pointX, pointRadius, _paintPoint);
      canvas.drawCircle(pointX, pointInnerRadius, _paintPointCenter);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    final old = oldDelegate as _ProgressoPainter;
    return old.progress != this.progress ||
        old.start != this.start ||
        old.progressColor != this.progressColor ||
        old.backgroundColor != this.backgroundColor ||
        old.progressStrokeWidth != this.progressStrokeWidth ||
        old.backgroundStrokeWidth != this.backgroundStrokeWidth ||
        old.progressStrokeCap != this.progressStrokeCap ||
        old.backgroundStrokeCap != this.backgroundStrokeCap ||
        old.pointColor != this.pointColor ||
        old.pointInnerColor != this.pointInnerColor ||
        old.pointRadius != this.pointRadius ||
        old.pointInnerRadius != this.pointInnerRadius ||
        old.points != this.points;
  }
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';

import 'package:progresso/progresso.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Progresso Demo App',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: HomePage(title: 'Progresso Demo App'),
    );
  }
}

class HomePage extends StatefulWidget {
  HomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  Stream<double> tick() {
    return Stream<double>.periodic(const Duration(milliseconds: 100), (x) {
      return x / 100;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
          child: Padding(
        padding: EdgeInsets.all(16.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text("Basic Progress"),
            SizedBox(height: 10),
            Progresso(progress: 0.5),
            SizedBox(height: 20),
            Text("Points along line"),
            SizedBox(height: 10),
            Progresso(progress: 0.5, points: [0.1, 0.2, 0.4]),
            SizedBox(height: 20),
            Text("non-zero start position"),
            SizedBox(height: 10),
            Progresso(start: 0.3, progress: 0.5),
            SizedBox(height: 20),
            Text("Custom colors"),
            SizedBox(height: 10),
            Progresso(
                progress: 0.5,
                progressColor: Colors.red,
                backgroundColor: Colors.lightBlue),
            SizedBox(height: 20),
            Text("Line widths"),
            SizedBox(height: 10),
            Progresso(
                progress: 0.5,
                progressStrokeWidth: 20,
                backgroundStrokeWidth: 20),
            SizedBox(height: 20),
            Text("point size"),
            SizedBox(height: 10),
            Progresso(
                progress: 0.5,
                pointInnerRadius: 0,
                pointRadius: 10,
                points: [0.1, 0.2, 0.4]),
            SizedBox(height: 20),
            Text("Line caps"),
            SizedBox(height: 10),
            Progresso(
                progress: 0.5,
                progressStrokeCap: StrokeCap.round,
                backgroundStrokeCap: StrokeCap.round),
            SizedBox(height: 20),
            Text("Point color"),
            SizedBox(height: 10),
            Progresso(
                progress: 0.5,
                points: [0.1, 0.2, 0.4],
                pointColor: Colors.red,
                pointInnerColor: Colors.black),
            SizedBox(height: 20),
            Text("stream progress"),
            SizedBox(height: 10),
            StreamBuilder(
                stream: tick(),
                builder: (_, AsyncSnapshot<double> snapshot) {
                  if (snapshot.hasData) {
                    return Progresso(progress: snapshot.data);
                  } else {
                    return Progresso();
                  }
                }),
            SizedBox(height: 20),
          ],
        ),
      )),
    );
  }
}
